consensus/state.go

```diff
@@ -1388,6 +1394,17 @@ func (cs *ConsensusState) addVote(vote *types.Vote, peerID p2p.ID) (added bool,
 						cs.LockedBlockParts = nil
 						cs.eventBus.PublishEventUnlock(cs.RoundStateEvent())
 					}
+					if ok && !blockID.IsZero() && !cs.ValidBlock.HashesTo(blockID.Hash) && vote.Round > cs.ValidRound {
+						// update valid value
+						if cs.ProposalBlock.HashesTo(blockID.Hash) {
+							cs.ValidRound = vote.Round
+							cs.ValidBlock = cs.ProposalBlock
+							cs.ValidBlockParts = cs.ProposalBlockParts
+						}
+						//TODO: We might want to update ValidBlock also in case we don't have that block yet,
```

Comment:
Oh good point. This might be a bit more complicated.

Well, I guess we just need to update the AddProposalBlockPart logic to check if we have ValidRound > 0 but no ValidBlock, then we know to set it with the complete ProposalBlock

Reply:
It's bit more tricky as ValidBlock might be (and probably will be) different than ProposalBlock. So we would need to extend gossiping to support sending BlockParts for two different blocks. That's why I am proposing addressing only the simplest scenario. We should still be correct with only this change, but the worst case termination time is O(N), where N is number of validators, instead of O(1) if we would be able to support ValidBlock mechanism fully. So it's performance issue, and we can address it after.